cmd/queuebench/main.go

```diff
@@ -0,0 +1,209 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package main
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+
+	apmqueue "github.com/elastic/apm-queue"
+	"github.com/elastic/apm-queue/kafka"
+)
+
+const namespace = "queuebench"
+
+func main() {
+	// NOTE: intercept any panic and terminate gracefully
+	// This allows using log.Panic in main which triggers
+	// deferred functions (whereas log.Fatal don't).
+	defer func() {
+		if r := recover(); r != nil {
+			log.Fatal(r)
+		}
+	}()
+
+	cfg := config{}
+	cfg.Parse()
+	log.Printf("parsed config: %+v\n", cfg)
+
+	log.Println("prep logger")
+	var logger *zap.Logger
+	var err error
+	if cfg.verbose {
+		logger, err = zap.NewDevelopment()
+		if err != nil {
+			log.Fatalf("cannot create zap logger: %s", err)
+		}
+	} else {
+		logger = zap.NewNop()
+	}
+
+	log.Println("prep MeterProvider")
+	rdr := sdkmetric.NewManualReader()
+	mp := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(rdr),
+	)
+
+	ctx := context.Background()
+	run := time.Now().Unix()
+	log.Printf("running bench run: %d", run)
+
+	bench := bench{
+		Brokers:         []string{cfg.broker},
+		ConsumerGroupID: fmt.Sprintf("queuebench-%d", run),
+		Logger:          logger,
+		Partitions:      cfg.partitions,
+		TopicNamespace:  namespace,
+		Topics: []apmqueue.Topic{
+			apmqueue.Topic(fmt.Sprintf("run-%d", run)),
+		},
+
+		mp: mp,
+		tp: trace.NewNoopTracerProvider(),
+	}
+
+	log.Println("running benchmark setup")
+	if err = bench.Setup(ctx); err != nil {
+		log.Panicf("benchmark setup failed: %s", err)
+	}
+	teardown := func() {
+		log.Println("running benchmark teardown")
+		if err := bench.Teardown(ctx); err != nil {
+			log.Panicf("benchmark teardown failed: %s", err)
+		}
+
+	}
+	defer teardown()
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		for range c {
+			teardown()
+		}
+	}()
```

Comment:
```suggestion
	ctx, stop := signal.NotifyContext(ctx, os.Interrupt)
	defer stop()
```

If we cancel the context on interrupt, then we can handle context cancellation to break the producing/consuming/wait loop. Then teardown can happen in one place.

Reply:
I refactored the logic based on your suggestion in [e1a0be2](https://github.com/elastic/apm-queue/pull/256/commits/e1a0be2a31c4166ba0cc9617da85e8752e550d4b)